### DISCUSSION TITLE
randomise left/right position of buttons

### DISCUSSION THREAD

**Discussion by mw719** [jspsych/jsPsych#1249] [Q&A] [answered]
Hi,

I'm still a beginner in jsPsych. Does anyone know how to randomise button positions (left/right) with the html-button-response plugin? I just couldn't get it work.

Thank you very much in advance!

**Comment by jodeleeuw** [comment, marked as answer]
Hi @mw719 ,

The button order is determined by the order in the `choices` array, so you can `shuffle()` the array to get a random order.

```js
var trial = {
	type: 'html-button-response',
	stimulus: 'Pick one.',
	choices: jsPsych.randomization.shuffle(['Option One', 'Option Two'])
}
```

The trouble with this is that you need to know the button order to know which one was pressed. You can solve this by saving the button order in the trial's data like this:

```js
var trial = {
	type: 'html-button-response',
	stimulus: 'Pick one.',
	choices: jsPsych.randomization.shuffle(['Option One', 'Option Two']),
	on_start: function(params){
		params.data.button_order = JSON.stringify(params.choices);
	}
}
```

The above method uses the `on_start` event to access the `choices` parameter and save it into the `data` parameter.

You could then save the label of the button clicked in the data when the trial finishes:

```js
var trial = {
	type: 'html-button-response',
	stimulus: 'Pick one.',
	choices: jsPsych.randomization.shuffle(['Option One', 'Option Two']),
	on_start: function(params){
		params.data.button_order = JSON.stringify(params.choices);
	},
	on_finish: function(data){
		var button_order = JSON.parse(data.button_order);
		data.button_pressed_label = button_order[data.button_pressed]
	}
}
```

**Reply by mw719** [reply]
Hi @jodeleeuw,

thanks a lot for you help! This code interpreted my choice variables as the trials, but I wanted to have other variables (a whole list) to be randomised. When implementing jsPsych.randomization.shuffle() below on_start, it worked. Thank you so much again! One last question: what is exactly the meaning of JSON.stringify and JSON.parse? Aren't the button presses saved without it as well?

**Reply by jodeleeuw** [reply]
`JSON.stringify()` converts the array to a string representation of the array. `JSON.parse()` reverts the string back to the array. It's considered best practice in the current version of jsPsych to only store "primitive" data types in the data. So whenever we need to save `objects` or `arrays` we save them as their JSON-encoded string instead.

**Reply by mw719** [reply]
okay, that makes sense! Thank you so much for your help, @jodeleeuw! :)
